serde_core/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+    pub use self::core::ffi::CStr;
+    #[cfg(feature = "std")]
+    pub use std::ffi::CStr;
+
+    #[cfg(all(not(no_core_cstr), feature = "alloc", not(feature = "std")))]
+    pub use alloc::ffi::CString;
+    #[cfg(feature = "std")]
+    pub use std::ffi::CString;
+
+    #[cfg(feature = "std")]
+    pub use std::{error, net};
+
+    #[cfg(feature = "std")]
+    pub use std::collections::{HashMap, HashSet};
+    #[cfg(feature = "std")]
+    pub use std::ffi::{OsStr, OsString};
+    #[cfg(feature = "std")]
+    pub use std::hash::{BuildHasher, Hash};
+    #[cfg(feature = "std")]
+    pub use std::io::Write;
+    #[cfg(feature = "std")]
+    pub use std::path::{Path, PathBuf};
+    #[cfg(feature = "std")]
+    pub use std::sync::{Mutex, RwLock};
+    #[cfg(feature = "std")]
+    pub use std::time::{SystemTime, UNIX_EPOCH};
+
+    #[cfg(all(feature = "std", no_target_has_atomic, not(no_std_atomic)))]
+    pub use std::sync::atomic::{
+        AtomicBool, AtomicI16, AtomicI32, AtomicI8, AtomicIsize, AtomicU16, AtomicU32, AtomicU8,
+        AtomicUsize, Ordering,
+    };
+    #[cfg(all(feature = "std", no_target_has_atomic, not(no_std_atomic64)))]
+    pub use std::sync::atomic::{AtomicI64, AtomicU64};
+
+    #[cfg(all(feature = "std", not(no_target_has_atomic)))]
+    pub use std::sync::atomic::Ordering;
+    #[cfg(all(feature = "std", not(no_target_has_atomic), target_has_atomic = "8"))]
+    pub use std::sync::atomic::{AtomicBool, AtomicI8, AtomicU8};
+    #[cfg(all(feature = "std", not(no_target_has_atomic), target_has_atomic = "16"))]
+    pub use std::sync::atomic::{AtomicI16, AtomicU16};
+    #[cfg(all(feature = "std", not(no_target_has_atomic), target_has_atomic = "32"))]
+    pub use std::sync::atomic::{AtomicI32, AtomicU32};
+    #[cfg(all(feature = "std", not(no_target_has_atomic), target_has_atomic = "64"))]
+    pub use std::sync::atomic::{AtomicI64, AtomicU64};
+    #[cfg(all(feature = "std", not(no_target_has_atomic), target_has_atomic = "ptr"))]
+    pub use std::sync::atomic::{AtomicIsize, AtomicUsize};
+}
+
+// None of this crate's error handling needs the `From::from` error conversion
+// performed implicitly by the `?` operator or the standard library's `try!`
+// macro. This simplified macro gives a 5.5% improvement in compile time
+// compared to standard `try!`, and 9% improvement compared to `?`.
+macro_rules! tri {
+    ($expr:expr) => {
+        match $expr {
+            Ok(val) => val,
+            Err(err) => return Err(err),
+        }
+    };
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+#[macro_use]
+mod macros;
+
+#[macro_use]
+mod integer128;
+
+pub mod de;
+pub mod ser;
+
+#[doc(inline)]
+pub use crate::de::{Deserialize, Deserializer};
+#[doc(inline)]
+pub use crate::ser::{Serialize, Serializer};
+
+// Used by generated code and doc tests. Not public API.
+#[doc(hidden)]
+#[path = "private/mod.rs"]
+pub mod __private;
```

Comment:
It looks like to keep `private` in `serde`, we'd have to figure out what to do about `__deserialize_content` in `Deserializer` trait. `private::de` is a substantial contributor to `private`'s line count, so I guess having some of `private` in `serde` and then rest in `serde_core` wouldn't quite cut it.
Instead, I've added a new feature to serde_core (`private`) that enables compilation of `serde_core::private::{de, ser}` module (kinda like `no_serde_derive` rustc cfg). `serde_core::private` is then enabled with `serde::derive`.

I've also updated other benchmarks, as one in PR description was taken on a relatively old machine I don't have access to right now.
Master (derive=true, 3.8s)
![image](https://github.com/serde-rs/serde/assets/24362066/c1580d28-eb18-48ec-a20d-ded8f8740aee)
This branch, commit 49d028d, derive=true, 2.4s:
![image](https://github.com/serde-rs/serde/assets/24362066/7dc2af5a-efcc-4864-8c87-79f9b4476b72)
This branch, patch 
[private_feature.patch](https://github.com/serde-rs/serde/files/14448278/private_feature.patch), derive=false:
![image](https://github.com/serde-rs/serde/assets/24362066/671e4f53-25aa-4cad-9b33-d4021f412e5e)

So all in all, getting rid of `private` in serde_core would net us about 200ms out of 1400ms on M1 Max.

Reply:
My concern about having `private` in `serde_core` is that it requires continued strict coupling between `serde_core` and `serde_derive`.  For example, I suspect this would make having a `serde_derive` v2 more difficult because you couldn't run it in parallel to `serde_derive` v1 without much finagling.

Reply:
I understood it as "moving private to serde" is a net gain, so that's what I meant with "let's do that":

Let's move it to serde and out of serde_core entirely

Reply:
Hey,
Just to give an update, I've encountered an issue with non-public API of Deserialize (`__deserialize_content`); should it still be a part of `serde_core`? If so, `Content` type would have to be in `serde_core` as well, which isn't ideal due to [reasons outlined by dtolnay](https://github.com/serde-rs/serde/pull/2608#pullrequestreview-1899765113).  I've tried numerous angles to remove `__deserialize_content` from `Deserialize` (and somehow move it's implementation into `serde`), without luck.

Reply:
I guess we'll have to revert https://github.com/serde-rs/serde/pull/2148 for now and reopen the issue it closed

Reply:
Alrighty, `private` has been moved into `serde`.